### DISCUSSION TITLE
Conditional Classes in React

### DISCUSSION THREAD

**Discussion by parkerduff** [tailwindlabs/tailwindcss#9009] [Help]
I am building a component library using react, tailwindcss, and typescript. This library will include all of the base components that we will use in our applications.
**What do people think the best approach is for building a base component that can have dozens of different styles depending on the combination of props?**

To provide an example, for our button component, we would like it to accept props of size, variant, and color. We also want to provide an optional className override that takes precedence over any of the classes specified in the props.

For instance, if I use this button, passing in the following props: color=‘primary’ size=‘small’ className=‘bg-red-500 text-lg’ then all of the styles of size=‘small’ color=‘primary’ and default variant should be applied except for what is overriden by the classes provided in className (bg-red-500 and text-xl).

Here's the example code:

```
import { ButtonHTMLAttributes, PureComponent } from 'react';
import { twMerge } from 'tailwind-merge';

type ButtonVariant = 'contained' | 'borderless' | 'outlined';
export type ButtonSize = 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge';
export type ButtonColor = 'primary' | 'light' | 'dark' | 'success' | 'error';
export type ContainedButtonColor =
  | 'primary'
  | 'gradient'
  | 'light'
  | 'dark'
  | 'success'
  | 'error';

interface CommonProps
  extends Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'color'> {
  size?: ButtonSize;
  children?: string;
}

export type ContainedButtonProps = CommonProps & {
  variant?: 'contained';
  color?: ContainedButtonColor;
};

export type BorderlessOutlinedButtonProps = CommonProps & {
  variant?: 'borderless' | 'outlined';
  color?: ButtonColor;
};

export type ButtonProps = ContainedButtonProps | BorderlessOutlinedButtonProps;

const styles = {
  xsmall: 'text-xs font-medium px-3 py-1.5',
  small: 'text-sm leading-4 font-medium px-3.5 py-2',
  medium: 'text-sm font-medium px-4 py-2',
  large: 'text-base font-medium px-5 py-2',
  xlarge: 'text-base font-medium px-6 py-3',
  contained: {
    base: '',
    primary: 'bg-blue-600 text-white',
    gradient: 'bg-gradient-to-r from-cyan-500 to-blue-500 text-white',
    light: 'bg-white text-neutral-900',
    dark: 'bg-neutral-900 text-white',
    success: 'bg-green-400 text-white',
    error: 'bg-red-300 text-white',
  },
  borderless: {
    base: 'shadow-none bg-transparent',
    primary: 'text-blue-600',
    light: 'text-white',
    dark: 'text-neutral-900',
    success: 'text-green-400',
    error: 'text-red-300',
  },
  outlined: {
    base: 'bg-white border border-solid',
    primary: 'text-blue-600 border-blue-600',
    light: 'text-neutral-900 border-neutral-900',
    dark: 'bg-neutral-900 text-white',
    success: 'text-green-400 border-green-400',
    error: 'text-red-300 border-red-300',
  },
};

export class Button extends PureComponent<ButtonProps> {
  override render(): JSX.Element {
    const {
      type,
      children,
      className,
      variant = 'contained',
      size = 'medium',
      color = 'primary',
      ...rest
    } = this.props;

    return (
      <button
        type={type ?? 'button'}
        className={twMerge(
          'rounded-full shadow-sm',
          styles[variant]['base'],
          styles[size],
          color === 'gradient'
            ? styles['contained'][color]
            : styles[variant][color],
          className
        )}
        {...rest}
      >
        {children}
      </button>
    );
  }
}

export default Button;
```

`<Button color='primary' size='small' className='bg-red-500 text-lg'>Overridden small primary button</Button>
`

In this example, we have the different styles stored in a JavaScript object. We then use the npm package tailwind-merge so that precedence goes from right to left. 
**Is there a better way of handling this situation?** Given the complexity of different prop combinations, having all of the classes inside the actual component is not feasible. We also looked into breaking out the classes into an scss module for each component using the @apply directive (though this created some problems with the className override). Performance of these base components is very important, and we don’t want to unnecessarily increase the javascript bundle size by storing all of the conditional classes in JS constants.

**Comment by wongjn** [comment]
Will you be shipping the generated Tailwind CSS with your components? Do you expect consumers of your components to use Tailwind classes in `className`? Because if you do, you would need to include every single Tailwind class (because you may not know what they will use) and this would really increase your CSS file size, or you would need to document which Tailwind classes are available for consumers to use. I don't think there's anyway around having classes inside the component; the classes need to live somewhere!

**Reply by parkerduff** [reply]
Yes, we expect that consumers (in our case, the apps that use the base component library) would be able to use all of the tailwind utility classes in an optional className prop (for some of the components).
I think I understand what you are getting at. To clarify my original question: do you think that putting the tailwind classes found in the styles constant _instead_ in a scss submodule is a better option than leaving them in a JavaScript const? Or is the example above the only practical way to handle this situation?

**Reply by wongjn** [reply]
I think it works as-is.

**Comment by meness** [comment]
Check [classnames](https://www.npmjs.com/package/classnames) out.
```jsx
className={classNames({"text-black": isBlack, "text-white": !isBlack})}
```

**Reply by parkerduff** [reply]
Thanks for the reply! We are probably going to use the classnames package in certain situations. That's the one we were originally using here before we realized we needed to pass in multiple conflicting classes to classnames and predictably use the one we want. This is where [tailwind-merge](https://www.npmjs.com/package/tailwind-merge) comes into play for us.
